cli/app/app.go

```diff
@@ -509,36 +509,70 @@ func configEnvs(name string, service ServiceConfig) []api.EnvVar {
 func configVolumes(service ServiceConfig) ([]api.VolumeMount, []api.Volume) {
 	volumesMount := []api.VolumeMount{}
 	volumes := []api.Volume{}
+	volumeSource := api.VolumeSource{}
 	for _, volume := range service.Volumes {
-		character := ":"
-		if strings.Contains(volume, character) {
-			containerDir := volume[strings.Index(volume, character)+1:]
-			containerDir = strings.TrimSpace(containerDir)
-
-			// check if ro/rw mode is defined
-			readonly := true
-			if strings.Index(volume, character) != strings.LastIndex(volume, character) {
-				mode := volume[strings.LastIndex(volume, character)+1:]
-				if strings.Compare(mode, "rw") == 0 {
-					readonly = false
-				}
-				containerDir = containerDir[0:strings.Index(containerDir, character)]
-			}
-
-			// volumeName = random string of 20 chars
-			volumeName := RandStringBytes(20)
+		name, host, container, mode, err := parseVolume(volume)
+		if err != nil {
+			logrus.Warningf("Failed to configure container volume: %v", err)
+			continue
+		}
 
-			volumesMount = append(volumesMount, api.VolumeMount{Name: volumeName, ReadOnly: readonly, MountPath: containerDir})
+		// volumeName = random string of 20 chars
+		if len(name) == 0 {
```

Comment:
I didn't assume that volume name is specified (even for v2, volume name isn't required), and if it's not, just use the current random volume name.